### PR TITLE
Fire events when a player connects or disconnects

### PR DESCRIPTION
**Pull request by SquidDev**
We provide connected and disconnected events for the client (when they have successfully connected to the server) and server (whenever a new client connects).

There are some caveats with this implementation, namely that the client-side events are fired on the network thread. It may be worth investigating if there are better places the hooks could be placed.

### CODE REVIEW

**Comment by modmuss50** [comment]
I think this looks good, will also lead into validating registry's and mods on connect

File: src/main/java/net/fabricmc/fabric/api/events/network/ConnectionEvent.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.events.network;
+
+import net.fabricmc.fabric.util.HandlerArray;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.text.TextComponent;
+
+public class ConnectionEvent {
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/api/events/network/ConnectionEvent.java]: Can we split this into ClientConnectionEvent and ServerConnectionEvent, with clearer event names CONNECTED and DISCONNECTED?

Comment (SquidDev) [reply src/main/java/net/fabricmc/fabric/api/events/network/ConnectionEvent.java]: My worry about this, and the reason I went for the longer names, is that I'm not actually sure how much clearer the shorter names are. I think there's a potential for ambiguity about whether `ServerConnectionEvent` is fired on the server or clients (when connecting to a server).

I guess one has the same issue with packets (does client refer to the sender or receiver), so we have a pattern already.

**Comment by Boundarybreaker** [comment]
This needs to be updated for the new event system.

**Comment by SquidDev** [comment]
@Boundarybreaker Sorry, entirely missed this comment. I'll look into this later today - I don't know if anyone else has thoughts on the event names though?

**Comment by grondag** [comment]
This appears to be redundant given that we have more recent PR in #247

**Pull request closed without merging**